src/rstudioapi.ts

```diff
@@ -0,0 +1,434 @@
+import {
+  window, TextEditor, TextDocument, Uri,
+  workspace, WorkspaceEdit, Position, Range, Selection,
+  QuickPickItem, QuickPickOptions, ViewColumn
+} from 'vscode';
+import { sessionDir, sessionDirectoryExists, writeResponse, writeSuccessResponse } from './session';
+import fs = require('fs-extra');
+import path = require('path');
+import { chooseTerminal, runTextInTerm, restartRTerminal } from './rTerminal';
+import { config } from './util';
+
+let lastActiveTextEditor: TextEditor;
+
+export async function dispatchRStudioAPICall(action: string, args: any, sd: string) {
+
+  switch (action) {
+case 'active_editor_context': {
+                    await writeResponse(await activeEditorContext(), sd);
+                    break;
+                }
+                case 'insert_or_modify_text': {
+                    await insertOrModifyText(args.query, args.id);
+                    await writeSuccessResponse(sd);
+                    break;
+                }
+                case 'replace_text_in_current_selection': {
+                    await replaceTextInCurrentSelection(args.text, args.id);
+                    await writeSuccessResponse(sd);
+                    break;
+                }
+                case 'show_dialog': {
+                    await showDialog(args.message);
+                    await writeSuccessResponse(sd);
+                    break;
+                }
+                case 'navigate_to_file': {
+                    await navigateToFile(args.file, args.line, args.column);
+                    await writeSuccessResponse(sd);
+                    break;
+                }
+                case 'set_selection_ranges': {
+                    await setSelections(args.ranges, args.id);
+                    await writeSuccessResponse(sd);
+                    break;
+                }
+                case 'document_save': {
+                    await documentSave(args.id);
+                    await writeSuccessResponse(sd);
+                    break;
+                }
+                case 'document_save_all': {
+                    await documentSaveAll();
+                    await writeSuccessResponse(sd);
+                    break;
+                }
+                case 'get_project_path': {
+                    await writeResponse(await projectPath(), sd);
```

Comment:
`projectPath` does not return a `Promise`. Not sure why the linter does not complain.

Reply:
my linter was actually flagging locally but not in the 'Problems' pane for some reason. Weird. Fixed!